File: Searching and Sorting/Cocktail Sort/cocktailSort.py
def cocktailSort(unsorted):
    swapped = True
    start = 0 
    end = len(unsorted) -1
    while swapped:
        swapped = False
        # forward pass
        for i in range(start, end):
            if unsorted[i] > unsorted[i+1]:
                swapped = True
                unsorted[i], unsorted[i+1] = unsorted[i+1], unsorted[i]

        end -= 1

        # backward pass
        for i in range(end-1, start-1, -1):
            if unsorted[i] > unsorted[i+1]:
                swapped = True
                unsorted[i], unsorted[i+1] = unsorted[i+1], unsorted[i]
        
        start += 1
        print(unsorted)


def test():
    arr = [3,4,8,76,45,3,1,0,5,-34, 23]
    cocktailSort(arr)


if __name__ == "__main__":
    test()

File: Searching and Sorting/Cocktail Sort/cocktailSort.c
#include<stdio.h>
#include<stdlib.h>

typedef enum{false, true} bool;

void cocktailSort(int unsorted[], int length);
void swap(int *a, int *b);
void printArray(int arr[], int length);

int main(void){
    int arr[] = {65,-3,23,4,0};
    int length = 5;
    cocktailSort(arr, length);
    printArray(arr, length);
    return 0;
}


void cocktailSort(int unsorted[], int length){
    bool swapped = true;
    int start = 0;
    int end = length -1;
    while (swapped){
        swapped = false;
        int i;
        for(i=start; i<end; i++){
            if (unsorted[i] > unsorted[i+1]){
                swap(&unsorted[i], &unsorted[i+1]);
                swapped = true;
            }
        }
        end --;
        for(i=end; i>start-1;i--){
            if(unsorted[i] > unsorted[i+1]){
                swap(&unsorted[i], &unsorted[i+1]);
                swapped = true;
                
            }
        }
    }
}

void swap(int *a, int *b){
    int tmp = *b;
    *b = *a;
    *a = tmp;
}

void printArray(int arr[], int length){
    int i;
    printf("[");
    for(i=0;i<length;i++){
        if((i+1) % 7 == 0){
            printf("\n ");
        }
        printf("%d%s", arr[i], (i==length-1 ? "]\n": ", "));
    }
}
